torch/csrc/jit/operator.h

```diff
@@ -79,15 +82,32 @@ struct TORCH_API Operator {
                 {},
                 /*is_vararg*/ true,
                 /*is_varret*/ true),
-            std::move(op_creator)) {}
+            std::move(op_creator),
+            std::move(options)) {}
 
-  Operator(FunctionSchema schema, Operation op)
+  Operator(FunctionSchema schema, Operation op, Options options = Options())
       : schema_(std::make_shared<FunctionSchema>(std::move(schema))),
-        op_(std::make_shared<Operation>(std::move(op))) {}
+        op_(std::make_shared<Operation>(std::move(op))),
+        options_(std::move(options)) {}
 
-  Operator(const std::string& schema, Operation op)
+  Operator(const std::string& schema, Operation op, Options options = Options())
       : schema_string_(schema),
-        op_(std::make_shared<Operation>(std::move(op))) {}
+        op_(std::make_shared<Operation>(std::move(op))),
+        options_(std::move(options)) {}
+
+  struct Options {
+    Options() {};
+    using AliasAnalysisFunction = void (AliasDb::*)(Node*);
```

Comment:
Making AliasAnalysisFunction a pointer rather than an enum doesn't buy us anything here, but it does strongly couple it to the alias_analysis.h header. How about just an enum in a separate header:

```
enum AliasAnalysisKind {
   CREATOR,
   EXTRACTOR, 
}

struct Options {
};
```

We can add more wildly customizable versions by patching options when necessary.

Reply:
I agree. The `analyzeFoo` functions in alias analysis are not really designed for outside consumption and are probably not clear to users. I think we could have an enum that covers the most common cases:
- Purely functional (no aliases, no mutation on inputs).
- View (output aliases input)
- Free for all (all inputs may be written to, all outputs may alias any input)
- Look at the schema (reads a user-provided schema, allows people to do whatever fancy thing).

Reply:
>reads a user-provided schema
@suo Not sure I follow how this would work

Reply:
To be clear, you don't have to put all of those options in this PR (just wanted to enumerate some common cases we'll want to add later).

For "Look at the schema"—people defining a custom op can pass a JIT schema if they want; so we can instruct the alias analysis pass to use that schema to populate alias info (similar to how we do with builtin ops).